### DISCUSSION TITLE
Analyzer: Warn when `ThrowsAsync` is not being awaited

### DISCUSSION THREAD

**Discussion by leonardochaia** [xunit/xunit#2344] [Ideas]
Hey all!

We've been using `xUnit` on .NET 5 for a while now and we are loving it.
Some thing that I have noticed on junior devs that are just getting started, is that they forget to `await` the `Assert.ThrowsAsync` method.

For example:

```csharp
[Fact]
public void ThrowsAsync_ShouldWork()
{
    Task Callback() => throw new Exception();
    Assert.ThrowsAsync<InvalidOperationException>(Callback);
}
``` 

Is there a chance that the xUnit analyzer can warn about this issues?
The example is giving a false positive.

If we await, the test fails:
```csharp
[Fact]
public async Task ThrowsAsync_ShouldWork()
{
    Task Callback() => throw new Exception();
    await Assert.ThrowsAsync<InvalidOperationException>(Callback);
}
```

So, I think it would be nice if the analyzer would warn about this cases, since the included .NET analyzers can't really know that you don't wanna do anything with the resulting Task.

**Comment by bradwilson** [comment]
In general we've tended to shy away from analyzers which are generally useful outside of xUnit.net (like a general failure to `await` anything which is async), but in this case I tend to agree. We should pit-of-success here and have analyzers for all async assertions to ensure that you're either awaiting them, or grabbing the task to use elsewhere, so I'm going to convert this to an issue.

**Reply by bradwilson** [reply]
Marking this question as resolved because of the issue conversion: https://github.com/xunit/xunit/issues/2346

**Reply by leonardochaia** [reply]
> We should pit-of-success here and have analyzers for all async assertions to ensure that you're either awaiting them, or grabbing the task to use elsewhere

I definitively agree. You must do something with the returning `Task` or else its pointless and definitively a code smell.

Thank you so much @bradwilson , my team and I appreciate it!!
